File: packages/redaction/src/pipeline.ts
/**
 * The 4-layer pre-persistence pipeline (SPEC §5):
 *
 *   capture (layer 1, C4 structural suppression — see capture.ts)
 *     → layer 2: category blocklist (C5) — a hit records NOTHING
 *     → layer 3: regex battery (always) + NER sidecar — FAIL-CLOSED:
 *       if the sidecar is unavailable the pipeline halts; nothing persists.
 *       The battery still having run is not grounds to degrade.
 *     → persistence (SQLCipher in the daemon; injected sink here)
 *     → layer 4: end-of-day review (review.ts) — a right, not a chore.
 *
 * Nothing reaches the sink except a fully-redacted ObserverEvent. There is no
 * code path that persists a RawCaptureEvent.
 */
import { applyBattery, classifyValue } from './battery.js';
import { blockedCategoryFor, EMPTY_EXCLUSIONS, type UserExclusions } from './blocklist.js';
import { NerUnavailableError, type NerClient } from './ner.js';
import { hostOf, scrubUrl } from './url.js';
import type { ObserverEvent, ProcessResult, RawCaptureEvent } from './types.js';

export interface PersistSink {
  append(event: ObserverEvent): Promise<void>;
}

export interface PipelineOptions {
  ner: NerClient;
  sink: PersistSink;
  exclusions?: UserExclusions;
  /** Called when the pipeline halts so the daemon can suspend capture. */
  onHalt?: (reason: string) => void;
}

export class RedactionPipeline {
  private readonly ner: NerClient;
  private readonly sink: PersistSink;
  private exclusions: UserExclusions;
  private readonly onHalt: ((reason: string) => void) | undefined;
  private haltedReason: string | null = null;

  constructor(opts: PipelineOptions) {
    this.ner = opts.ner;
    this.sink = opts.sink;
    this.exclusions = opts.exclusions ?? EMPTY_EXCLUSIONS;
    this.onHalt = opts.onHalt;
  }

  get halted(): boolean {
    return this.haltedReason !== null;
  }

  /** Layer-4 exclusions feed back into layer 2 for the rest of the study. */
  addExclusions(more: Partial<UserExclusions>): void {
    this.exclusions = {
      hosts: [...this.exclusions.hosts, ...(more.hosts ?? [])],
      bundleIds: [...this.exclusions.bundleIds, ...(more.bundleIds ?? [])],
      appNames: [...this.exclusions.appNames, ...(more.appNames ?? [])],
    };
  }

  /** Re-arm after the sidecar supervisor reports healthy again. */
  resume(): void {
    this.haltedReason = null;
  }

  async process(raw: RawCaptureEvent): Promise<ProcessResult> {
    // Fail-closed: once halted, nothing persists until the supervisor resumes us.
    if (this.haltedReason !== null) {
      return { action: 'halted_ner_unavailable' };
    }

    // Layer 2 — category blocklist (C5), before any redaction work.
    const category = blockedCategoryFor(
      raw.window,
      { bundleId: raw.app.bundleId, name: raw.app.name },
      hostOf(raw.url),
      this.exclusions,
    );
    if (category !== null) {
      return { action: 'blocked_category', category };
    }

    // Layer 3 — battery first (sync, always), then NER (fail-closed).
    const titlePass = applyBattery(raw.window.title);
    const labelPass = raw.ax && !raw.ax.secureSuppressed ? applyBattery(raw.ax.label) : null;

    let titleRedacted: string;
    let labelRedacted: string | null;
    let nerRules: string[] = [];
    try {
      const titleNer = await this.ner.redact(titlePass.text);
      titleRedacted = titleNer.redacted;
      nerRules = titleNer.rulesHit;
      if (labelPass !== null) {
        const labelNer = await this.ner.redact(labelPass.text);
        labelRedacted = labelNer.redacted;
        nerRules = [...nerRules, ...labelNer.rulesHit];
      } else {
        labelRedacted = null;
      }
    } catch (err) {
      if (err instanceof NerUnavailableError) {
        this.haltedReason = 'ner_unavailable';
        this.onHalt?.('ner_unavailable');
        return { action: 'halted_ner_unavailable' };
      }
      throw err;
    }

    // Values are classified, then discarded — the schema has no value field.
    const valueClass = raw.ax
      ? raw.ax.secureSuppressed
        ? ('none' as const)
        : classifyValue(raw.ax.value)
      : ('none' as const);

    const rulesHit = [...new Set([...titlePass.rulesHit, ...(labelPass?.rulesHit ?? []), ...nerRules])];

    const event: ObserverEvent = {
      v: 1,
      id: raw.id,
      ts: raw.ts,
      session: raw.session,
      kind: raw.kind,
      app: { bundle_id: raw.app.bundleId, name: raw.app.name },
      window: { title_redacted: titleRedacted, id: raw.window.id },
      url: raw.url !== undefined ? scrubUrl(raw.url) : null,
      ax: raw.ax
        ? {
            role_path: raw.ax.rolePath,
            action: raw.ax.action,
            // C4: a suppressed field contributes a constant, never its label.
            label_redacted: raw.ax.secureSuppressed ? '{SECURE}' : (labelRedacted ?? ''),
            value_class: valueClass,
          }
        : null,
      input: raw.input
        ? { keys: raw.input.keys, clicks: raw.input.clicks, duration_ms: raw.input.durationMs }
        : null,
      frame_ref: raw.ax?.secureSuppressed ? null : (raw.frameRef ?? null),
      redaction: { rules_hit: rulesHit, review_state: 'auto' },
    };

    await this.sink.append(event);
    return { action: 'persisted', event };
  }
}

File: apps/desktop/src-tauri/crates/nibbin-redaction/src/pipeline.rs
//! The 4-layer pre-persistence pipeline — Rust twin of @nibbin/redaction's
//! pipeline.ts. Nothing reaches the sink except a fully-redacted
//! ObserverEvent; there is no code path that persists a RawCaptureEvent
//! (RawCaptureEvent does not even implement Serialize).

use crate::battery::{apply_battery, classify_value};
use crate::blocklist::{blocked_category_for, UserExclusions};
use crate::event::{
    AppRef, AxObservation, AxRef, ObserverEvent, RawCaptureEvent, RedactionMeta, ReviewState,
    ValueClass, WindowRef,
};
use crate::ner::{NerClient, NerError};
use crate::url_scrub::{host_of, scrub_url};

#[derive(Debug, PartialEq, Eq)]
pub enum ProcessOutcome {
    Persisted,
    BlockedCategory(String),
    HaltedNerUnavailable,
}

pub trait PersistSink {
    fn append(&mut self, event: &ObserverEvent) -> Result<(), anyhow::Error>;
}

pub struct RedactionPipeline<N: NerClient> {
    ner: N,
    exclusions: UserExclusions,
    halted_reason: Option<String>,
}

impl<N: NerClient> RedactionPipeline<N> {
    pub fn new(ner: N) -> Self {
        Self {
            ner,
            exclusions: UserExclusions::default(),
            halted_reason: None,
        }
    }

    pub fn halted(&self) -> bool {
        self.halted_reason.is_some()
    }

    /// Layer-4 exclusions feed back into layer 2 for the rest of the study.
    pub fn add_exclusions(&mut self, more: UserExclusions) {
        self.exclusions.hosts.extend(more.hosts);
        self.exclusions.bundle_ids.extend(more.bundle_ids);
        self.exclusions.app_names.extend(more.app_names);
    }

    /// Re-arm after the sidecar supervisor reports healthy again.
    pub fn resume(&mut self) {
        self.halted_reason = None;
    }

    pub fn process(
        &mut self,
        raw: &RawCaptureEvent,
        sink: &mut dyn PersistSink,
    ) -> Result<ProcessOutcome, anyhow::Error> {
        // Fail-closed: once halted, nothing persists until resumed.
        if self.halted_reason.is_some() {
            return Ok(ProcessOutcome::HaltedNerUnavailable);
        }

        // Layer 2 — category blocklist (C5), before any redaction work.
        let url_host = raw.url.as_deref().and_then(host_of);
        if let Some(category) = blocked_category_for(
            &raw.window_title,
            raw.window_category.as_deref(),
            &raw.app_bundle_id,
            &raw.app_name,
            url_host.as_deref(),
            &self.exclusions,
        ) {
            return Ok(ProcessOutcome::BlockedCategory(category));
        }

        // Layer 3 — battery first (always), then NER (fail-closed).
        let title_pass = apply_battery(&raw.window_title);
        let label_pass = match &raw.ax {
            Some(AxObservation::Plain { label, .. }) => Some(apply_battery(label)),
            _ => None,
        };

        let mut rules_hit: Vec<String> = title_pass.rules_hit.clone();
        if let Some(lp) = &label_pass {
            rules_hit.extend(lp.rules_hit.clone());
        }

        let (title_redacted, label_redacted) = {
            let title_ner = match self.ner.redact(&title_pass.text) {
                Ok(r) => r,
                Err(NerError::Unavailable(reason)) => {
                    self.halted_reason = Some(reason);
                    return Ok(ProcessOutcome::HaltedNerUnavailable);
                }
            };
            rules_hit.extend(title_ner.rules_hit);

            let label_redacted = match &label_pass {
                Some(lp) => match self.ner.redact(&lp.text) {
                    Ok(r) => {
                        rules_hit.extend(r.rules_hit);
                        Some(r.redacted)
                    }
                    Err(NerError::Unavailable(reason)) => {
                        self.halted_reason = Some(reason);
                        return Ok(ProcessOutcome::HaltedNerUnavailable);
                    }
                },
                None => None,
            };
            (title_ner.redacted, label_redacted)
        };

        rules_hit.sort();
        rules_hit.dedup();

        // Values are classified, then discarded — the schema has no value field.
        let ax = raw.ax.as_ref().map(|obs| match obs {
            AxObservation::Plain {
                role_path,
                action,
                value,
                ..
            } => AxRef {
                role_path: role_path.clone(),
                action: *action,
                label_redacted: label_redacted.clone().unwrap_or_default(),
                value_class: classify_value(value.as_deref()),
            },
            AxObservation::SecureSuppressed { role_path, action } => AxRef {
                role_path: role_path.clone(),
                action: *action,
                // C4: a suppressed field contributes a constant, never its label.
                label_redacted: "{SECURE}".to_string(),
                value_class: ValueClass::None,
            },
        });

        let secure = matches!(raw.ax, Some(AxObservation::SecureSuppressed { .. }));

        let event = ObserverEvent {
            v: 1,
            id: raw.id.clone(),
            ts: raw.ts.clone(),
            session: raw.session.clone(),
            kind: raw.kind,
            app: AppRef {
                bundle_id: raw.app_bundle_id.clone(),
                name: raw.app_name.clone(),
            },
            window: WindowRef {
                title_redacted,
                id: raw.window_id.clone(),
            },
            url: raw.url.as_deref().and_then(scrub_url),
            ax,
            input: raw.input.clone(),
            frame_ref: if secure { None } else { raw.frame_ref.clone() },
            redaction: RedactionMeta {
                rules_hit,
                review_state: ReviewState::Auto,
            },
        };

        sink.append(&event)?;
        Ok(ProcessOutcome::Persisted)
    }
}
